docassemble/ALKilnInThePlayground/data/questions/run_alkiln_tests.yml

```
@@ -769,77 +854,241 @@ back button label: Install a different version
+    """
+    Return the status, message, and output of `install_process`.
+    """
+
+    if not install_process:
+      return { "ok": False, "messages": "`install_process` was a falsy value", "result": None}
+
+    ok = True
+    messages = []
+    install_result = None
+
+    install_pid = None
+    returncode = None
+
+    try:
+      process_wait_start_time = current_datetime()
+
+      install_pid = install_process.pid
+      # Get `(stdout_data, stderr_data)`
+      install_result = install_process.communicate( timeout=max_run_time )
+      returncode = install_process.returncode
+
+    except subprocess.TimeoutExpired:
+      ok = False
+      message = f"""The ALKiln installation process with the ID { install_pid } ran for over { date_difference( starting=process_wait_start_time, ending=current_datetime() ).seconds } seconds. The maximum time allowed is { max_run_time } seconds."""
+      log( f"""🤕 ALKiln ERROR: { message }""" )
+      messages.append( message )
+
+    except Exception as install_wait_error:
+      ok = False
+      message = f"""Error while installing ALKiln:
+  { "\n".join( traceback.format_exception( install_wait_error ) ) }"""
+      log(f"""🤕 ALKiln ERROR: { message }""")
+      messages.append( message )
+
+    finally:
+
+      # Force process to terminate if needed
+      try:
+        if install_process and install_pid:
+          # We have to terminate this process ourselves because
+          # when `.communicate` runs out of time, the system doesn't
+          # complete the process itself. Low level interface.
+          import os
+          import signal
+          # SIGTERM is not enough
+          os.killpg(os.getpgid( install_pid ), signal.SIGKILL)
+          install_process.wait()
+
+        else:
+          message = f"The ALKiln installation process ended and the process ID is missing. We have no more information, sorry."
+          log(f"""🤕 ALKiln ERROR: { message }""")
```

Warning instead? Do we want to fail installation quietly? If not, should we instead raise an Exception? After all, no pid means things went poorly.